From f30ece21ff80b6a59713449358f36d69590224b8 Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Thu, 13 Dec 2012 22:32:22 +0000
Subject: [PATCH 175/795] MR1: MAPREDUCE-4860. DelegationTokenRenewal attempts to renew token even after a job is removed. (kkambatl via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1421582 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit f89fdb59ea132ba9e74bfb53cce788584d2e9639)

Reason: Fix failing tests
Ref: CDH-9041
Author: Karthik Kambatla
(cherry picked from commit 9c8b1ab453053b942c93aad2558af2805c4ee68c)
(cherry picked from commit 45ad5a02a3a2ba644b7513fadafff3fcbc05fef6)
---
 .../security/token/DelegationTokenRenewal.java     |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/src/mapred/org/apache/hadoop/mapreduce/security/token/DelegationTokenRenewal.java b/src/mapred/org/apache/hadoop/mapreduce/security/token/DelegationTokenRenewal.java
index dfd8bca..d49defd 100644
--- a/src/mapred/org/apache/hadoop/mapreduce/security/token/DelegationTokenRenewal.java
+++ b/src/mapred/org/apache/hadoop/mapreduce/security/token/DelegationTokenRenewal.java
@@ -202,11 +202,16 @@ public static synchronized void registerDelegationTokensForRenewal(
    */
   private static class RenewalTimerTask extends TimerTask {
     private DelegationTokenToRenew dttr;
+    private boolean cancelled = false;
     
     RenewalTimerTask(DelegationTokenToRenew t) {  dttr = t;  }
     
     @Override
-    public void run() {
+    public synchronized void run() {
+      if (cancelled) {
+        return;
+      }
+
       Token<?> token = dttr.token;
       try {
         // need to use doAs so that http can find the kerberos tgt
@@ -229,6 +234,12 @@ public Long run() throws Exception {
         removeFailedDelegationToken(dttr);
       }
     }
+
+    @Override
+    public synchronized boolean cancel() {
+      cancelled = true;
+      return super.cancel();
+    }
   }
   
   /**
-- 
1.7.0.4

